llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -188,54 +189,141 @@ int PostPipeliner::fit(MachineInstr *MI, int First, int Last, int II) {
   return -1;
 }
 
-void PostPipeliner::computeLoopCarriedParameters() {
-  // We schedule the first iteration, only using earliest. This updates
-  // earliest of the successors. Any successor in the second iteration
-  // represents a loop carried dependence, and we account for that by
-  // propagating its Earliest back to the first iteration
-  // Note that we don't have to clean the effects of this exploration,
-  // since the real scheduling will overwrite Cycle, and the ultimate Earliest
-  // will never be less than we compute here.
-
+void PostPipeliner::computeForward() {
+  // The forward order defines a topological sort, so we can compute
+  // Earliest and Ancestors in a single forward sweep
+  // Also initialize Latest to the last cycle in the last stage
+  const int Latest = -1;
   for (int K = 0; K < NInstr; K++) {
+    auto &Me = Info[K];
+    Me.Latest = Latest;
     SUnit &SU = DAG->SUnits[K];
-    const int Earliest = Info[K].Earliest;
-    scheduleNode(SU, Earliest);
+    for (auto &Dep : SU.Preds) {
+      if (Dep.getKind() != SDep::Data) {
+        continue;
+      }
+      int P = Dep.getSUnit()->NodeNum;
+      assert(P < K);
+      Me.Ancestors.insert(P);
+      auto &Pred = Info[P];
+      for (int Anc : Pred.Ancestors) {
+        Me.Ancestors.insert(Anc);
+      }
+    }
+    for (auto &Dep : SU.Succs) {
+      auto *Succ = Dep.getSUnit();
+      if (Succ->isBoundaryNode()) {
+        continue;
+      }
+      auto &SInfo = Info[Succ->NodeNum];
+      const int NewEarliest = Me.Earliest + Dep.getSignedLatency();
+      SInfo.Earliest = std::max(SInfo.Earliest, NewEarliest);
+    }
   }
+}
 
-  // Propagate Earliest upstream, initialize Latest
-  // Unrestricted: last cycle of last stage
-  const int Latest = -1;
+bool PostPipeliner::computeBackward() {
+  bool Changed = false;
+
+  auto AddOffspring = [&Changed](NodeInfo &Info, int E) {
+    if (Info.Offspring.insert(E).second) {
+      Changed = true;
+    }
+  };
+
+  // Traversing backwards will speed convergence a bit
+  for (int K = NInstr - 1; K >= 0; K--) {
```

Comment:
Nit: Again, something like `for (NodeInfo &Node : reverse(Info))` would be clearer that having to deal with indices.

Reply:
But reverse Info runs over all Info nodes (NTotalInstr) Moreover, indices connect SUnits and Info

Reply:
I think this just means we need helpers like `iterator_range<NodeInfo> first_nodes()`, and `SUnit getSUnit(NodeInfo &Node)`. Again, this is a nit, nothing blocking :) But I think this would hide away implementation details and this means I don't have to remember what `K` and `NInstr` mean.